opsimulator/indexer.go

```diff
@@ -0,0 +1,130 @@
+}
+
+func NewL1ToL2MessageIndexer(log log.Logger, storeManager *L1DepositStoreManager) *L1ToL2MessageIndexer {
+	tasksCtx, tasksCancel := context.WithCancel(context.Background())
+
+	return &L1ToL2MessageIndexer{
+		log:          log,
+		storeManager: storeManager,
+		eb:           EventBus.New(),
+		tasks: tasks.Group{
+			HandleCrit: func(err error) {
+				fmt.Printf("unhandled indexer error: %v\n", err)
+			},
+		},
+		tasksCtx:    tasksCtx,
+		tasksCancel: tasksCancel,
+	}
+}
+
+func (i *L1ToL2MessageIndexer) Start(ctx context.Context, client *ethclient.Client, l2Chain config.Chain) error {
+
+	i.l2Chain = l2Chain
+
+	i.tasks.Go(func() error {
+		depositTxCh := make(chan *types.DepositTx)
+		logCh := make(chan types.Log)
+
+		defer close(depositTxCh)
+		defer close(logCh)
+
+		channels := DepositChannels{
+			DepositTxCh: depositTxCh,
+			LogCh:       logCh,
+		}
+
+		portalAddress := common.Address(l2Chain.Config().L2Config.L1Addresses.OptimismPortalProxy)
+		sub, err := SubscribeDepositTx(i.tasksCtx, client, portalAddress, channels)
+
+		if err != nil {
+			return fmt.Errorf("failed to subscribe to deposit tx: %w", err)
+		}
+
+		chainID := i.l2Chain.Config().ChainID
+
+		for {
+			select {
+			case dep := <-depositTxCh:
+				log := <-logCh
+				if err := i.processEvent(dep, log, chainID); err != nil {
+					fmt.Printf("failed to process log: %v\n", err)
+				}
+
+			case <-i.tasksCtx.Done():
+				sub.Unsubscribe()
+			}
+		}
+	})
+
+	return nil
+}
+
+func (i *L1ToL2MessageIndexer) Stop(ctx context.Context) error {
+	i.tasksCancel()
+	return nil
+}
+
+func depositMessageInfoKey(destinationChainID uint64) string {
+	return fmt.Sprintf("DepositMessageKey:destination:%d", destinationChainID)
+}
+
+func (i *L1ToL2MessageIndexer) SubscribeDepositMessage(destinationChainID uint64, depositMessageChan chan<- *types.Transaction) (func(), error) {
+	return i.createSubscription(depositMessageInfoKey(destinationChainID), depositMessageChan)
+}
+
+func (i *L1ToL2MessageIndexer) createSubscription(key string, depositMessageChan chan<- *types.Transaction) (func(), error) {
+	handler := func(e *types.Transaction) {
+		depositMessageChan <- e
+	}
+
+	if err := i.eb.Subscribe(key, handler); err != nil {
+		return nil, fmt.Errorf("failed to create subscription %s: %w", key, err)
+	}
+
+	return func() {
+		_ = i.eb.Unsubscribe(key, handler)
+	}, nil
+}
+
+func (i *L1ToL2MessageIndexer) processEvent(dep *types.DepositTx, log types.Log, chainID uint64) error {
+
+	depTx := types.NewTx(dep)
+	i.log.Info("observed deposit event on L1", "hash", depTx.Hash().String(), "SourceHash", dep.SourceHash.String())
+
+	depositMessage := L1DepositMessage{
+		DepositTxn: dep,
+		DepositLog: log,
+	}
+
+	if err := i.storeManager.Set(dep.SourceHash, &depositMessage); err != nil {
+		i.log.Error("failed to store deposit tx to chain: %w", "chain.id", chainID, "err", err)
+		return err
+	}
+
+	i.eb.Publish(depositMessageInfoKey(chainID), depTx)
+	return nil
+}
```

Comment:
We should add a test at the indexer level in indexer_test.go. it should confirm that sending a tx gets correctly picked up, and the subscribe methods work correctly.